llzk_backend/src/template.rs

```diff
@@ -686,13 +727,15 @@ where
                                         )?
                                         .into(),
                                     )?;
+                                    let (lhs, rhs) = unify_constrain_eq_types(
+                                        fc,
+                                        codegen.location_from_meta(meta),
+                                        val_from_read,
+                                        *val,
+                                    )?;
                                     fc.append_op_no_result(
-                                        constrain::eq(
-                                            codegen.location_from_meta(meta),
-                                            val_from_read,
-                                            *val,
-                                        )
-                                        .into(),
+                                        constrain::eq(codegen.location_from_meta(meta), lhs, rhs)
+                                            .into(),
                                     )?;
                                     fc.block_ctx.set_named_value(var.clone(), *val)
```

Comment:
I noticed one more extra cast in `circom/tests/simple/read_from_output_with_felt_cast.circom` that should be avoidable by using `rhs` here instead of `val`